valheim-updater

```diff
@@ -5,39 +5,81 @@ CONFIG_DIRECTORY_PERMISSIONS=${CONFIG_DIRECTORY_PERMISSIONS:-755}
 WORLDS_DIRECTORY_PERMISSIONS=${WORLDS_DIRECTORY_PERMISSIONS:-755}
 WORLDS_FILE_PERMISSIONS=${WORLDS_FILE_PERMISSIONS:-644}
 UPDATE_INTERVAL=${UPDATE_INTERVAL:-900}
-STEAMCMD_ARGS=${STEAMCMD_ARGS-validate}
+STEAMCMD_ARGS=${STEAMCMD_ARGS}
+FORCE_VALIDATE=${FORCE_VALIDATE:-false}
+VALHEIM_PLUS_ENABLED=${VALHEIM_PLUS_ENABLED:-true}
 just_started=true
 next_update=$(date +%s)
 
 cd /opt/steamcmd
 main() {
     while :; do
         update_permissions
-        update
+        update_valheim
+        if [ $VALHEIM_PLUS_ENABLED = true ]; then
+            update_valheim_plus
+        fi
         next_update=$(($(date +%s)+$UPDATE_INTERVAL))
         while [ $(date +%s) -lt $next_update ]; do
             sleep 5
         done
     done
 }
 
-update() {
-    local logfile="$(mktemp)"
-    echo "Updating/Validating Valheim Server"
-    ./steamcmd.sh +login anonymous +force_install_dir /opt/valheim_dl +app_update 896660 $STEAMCMD_ARGS +quit
-    rsync -a --itemize-changes --delete --exclude server_exit.drp --exclude steamapps /opt/valheim_dl/ /opt/valheim | tee "$logfile"
-    grep '^[*>]' "$logfile" > /dev/null 2>&1
-    if [ $? -eq 0 ]; then
-        echo "Valheim Server was updated - restarting"
-        supervisorctl restart valheim-server
+install_valheim_plus() {
+    echo "Installing/Updating Valheim Plus"
+    wget -q -O /opt/valheimplus/UnixServer.zip https://github.com/nxPublic/ValheimPlus/releases/latest/download/UnixServer.zip && \
+    unzip -qo /opt/valheimplus/UnixServer.zip -x changelog.txt run_bepinex.sh -d /opt/valheimplus
+    if [ -f /opt/valheim/BepInEx/config/valheim_plus.cfg ]; then
+        cp -a /opt/valheim/BepInEx/config/valheim_plus.cfg /opt/valheimplus/BepInEx/config/valheim_plus.cfg
+    fi
+    chmod 666 /opt/valheimplus/BepInEx/config/valheim_plus.cfg
+    rsync -a --itemize-changes --exclude UnixServer.zip /opt/valheimplus/ /opt/valheim
+}
+
+update_valheim_plus() {
+    remote_md5="$(curl -sL https://github.com/nxPublic/ValheimPlus/releases/latest/download/UnixServer.zip | md5sum | cut -d ' ' -f 1)"
+    local_md5="$(md5sum /opt/valheimplus/UnixServer.zip | cut -d ' ' -f 1)"
+    if [ "$remote_md5" != "$local_md5" ]; then
+        install_valheim_plus
+    else
+        echo "Valheim Plus is already the latest version"
+    fi
+}
+
+install_valheim() {
+    ./steamcmd.sh +login anonymous +force_install_dir /opt/valheim +app_update 896660 $1 $STEAMCMD_ARGS +quit
```

Comment:
What happens to existing users who extended STEAMCMD_ARGS and already have validate plus whatever else they were using in here? Do we have to remove validate from their STEAMCMD_ARGS in order for this mod to work? Should we tell them in the README that validate in STEAMCMD_ARGS is not compatible with ValheimPlus mod?

Reply:
With regards to this, I think it is better to strip `-validate` from `STEAMCMD_ARGS` and just redirect that behavior to `FORCE_VALIDATE` instead. This has the benefit of making `-validate` a separate and intentional flag while not breaking current user `STEAMCMD_ARGS`. Although I don't believe for Valheim Server there are any additional arguments that can be passed to `steamcmd` as they don't have a publicly available beta branch